router/src/main/java/io/micronaut/web/router/version/VersionedRouterDecoratorConfiguration.java

```diff
@@ -0,0 +1,42 @@
+package io.micronaut.web.router.version;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.context.event.BeanCreatedEvent;
+import io.micronaut.context.event.BeanCreatedEventListener;
+import io.micronaut.web.router.Router;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Configuration to decorate {@link Router} with version matching logic.
+ */
+@Singleton
+@Requires(beans = RoutesVersioningConfiguration.class)
+public class VersionedRouterDecoratorConfiguration implements BeanCreatedEventListener<Router> {
```

Comment:
I'm not sure it is necessary to wrap the existing router. I think it would probably be better just to fold this behaviour into DefaultRouter

Reply:
Maybe instead of Implementing Router interface the VersionedRouter can extend DefaultRouter and override only single method `find(HttpRequest request)`?

Also by the documentation
`
The default {@link Router} implementation. This implementation does not perform any additional caching of route discovery
`

I think it will blow up the DefaultRouter in future.

Reply:
I'm not sure I understand the need for `VersionedRouter` 

As far as I can tell it doesn't do any caching? If it is necessary then it is better to defined `@Replaced(DefaultRouter.class)` on `VersionedRouter` rather than use a `BeanCreatedEventListener`

Reply:
The idea was to be flexible enough to support future realizations of `Router` (CachingRouter maybe etc).
The `VersionedRouter` with `BeanCreatedEventListener` decorates any existing router in `ApplicationContext` if versioning is enabled.
With `@Replaced` it will be hard to refactor in future (My opinion)

But if there are no plans to extend Router currenlty, I can use `@Replaced` or move the logic to `DefaultRouter`